Spigot-Server-Patches/0402-implement-optional-per-player-mob-spawns.patch

```diff
@@ -0,0 +1,244 @@
++                                                            mobsSpawned.add(entityinsentient);
++                                                        }
++
++                                                        if (mobsSpawned.size() >= maxSpawns){
++                                                            return mobsSpawned; // Paper
+                                                         }
++                                                        // Paper end
+                                                         // CraftBukkit end
+                                                         if (i >= entityinsentient.dC()) {
+-                                                            return;
++                                                            return mobsSpawned; // Paper
+                                                         }
+ 
+                                                         if (entityinsentient.c(i2)) {
+@@ -142,6 +157,7 @@ public final class SpawnerCreature {
+ 
+             }
+         }
++        return mobsSpawned; // Paper
+     }
+ 
+     @Nullable
+diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
+index 135ec94c..638708f1 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -17,6 +17,8 @@ import it.unimi.dsi.fastutil.objects.Object2IntMap;
+ import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+ import it.unimi.dsi.fastutil.objects.ObjectIterator;
+ import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
++import java.util.ArrayList; // Paper
++import java.util.HashMap; // Paper
+ import java.util.Iterator;
+ import java.util.List;
+ import java.util.Map;
+@@ -67,6 +69,7 @@ public class WorldServer extends World {
+     private boolean ticking;
+     @Nullable
+     private final MobSpawnerTrader mobSpawnerTrader;
++    private Map<EntityPlayer, Map<EnumCreatureType, Integer>> playerMobTypeMap; // Paper
+ 
+     // CraftBukkit start
+     private int tickPosition;
+@@ -835,6 +838,7 @@ public class WorldServer extends World {
+     }
+ 
+     public Object2IntMap<EnumCreatureType> l() {
++        List<Entity> filteredEntities = new ArrayList<>(); // Paper
+         Object2IntMap<EnumCreatureType> object2intmap = new Object2IntOpenHashMap();
+         ObjectIterator objectiterator = this.entitiesById.values().iterator();
+ 
+@@ -854,16 +858,49 @@ public class WorldServer extends World {
+                 EnumCreatureType enumcreaturetype = entity.getEntityType().d();
+ 
+                 if (enumcreaturetype != EnumCreatureType.MISC && this.getChunkProvider().a(entity)) {
++                    filteredEntities.add(entity); // Paper
+                     object2intmap.computeInt(enumcreaturetype, (enumcreaturetype1, integer) -> {
+                         return 1 + (integer == null ? 0 : integer);
```

Comment:
This is a Object2IntMap, so integer won't be null, but using the default value 0.

Reply:
Not sure what you are referring to, but lines 196-197 are mojang code